File: ml-benchmark/autolifter/single-pass/min.f
Config ExtraGrammar = "AutoLifter";

Inductive List = cons {Int, List} | nil Unit;

single_pass = \v: List -> Int.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

inf = 100;

/*User provided programs*/

min = fix (\f: List -> Int. \xs: List.
  match xs with
    nil _ -> inf
  | cons {h, t} ->
    let res = f t in if < res h then res else h
  end
);

main = single_pass min;

File: ml-benchmark/autolifter/single-pass/min.ml
type list_ = Cons of int * list_ | Nil of unit

let single_pass v =
  let rec run xs =
    match xs with
    | Nil _ -> xs
    | Cons (h, t) -> Cons (h, run t)
  in
  fun xs -> v (run xs)

let inf = 100

let rec min xs =
  match xs with
  | Nil _ -> inf
  | Cons (h, t) ->
      let res = min t in
      if res < h then res else h

let program = single_pass min